File: node/lib/context.js
'use strict';

// 请求级通用字段（Node 版 sdkctx）。用 AsyncLocalStorage 实现线程/并发隔离，
// 语义同 Python contextvars（见 spec/common-fields.md）。
//
// 框架适配器在入口用 bindRequest 包裹请求处理，SDK log/metrics 读取当前
// 上下文；未绑定则回退部署级默认。

const { AsyncLocalStorage } = require('async_hooks');

const storage = new AsyncLocalStorage();

function current() {
  return storage.getStore() || {};
}

function community() {
  return current().community || null;
}

function requestId() {
  return current().requestId || null;
}

function traceId() {
  return current().traceId || null;
}

// bindRequest(store, fn)：在 store（{community?, requestId?, traceId?}）内执行 fn。
// 会与已有上下文合并（缺省字段继承外层）。
function bindRequest(fields, fn) {
  const prev = storage.getStore() || {};
  const merged = {
    community: fields.community !== undefined ? fields.community : prev.community,
    requestId: fields.requestId !== undefined ? fields.requestId : prev.requestId,
    traceId: fields.traceId !== undefined ? fields.traceId : prev.traceId,
  };
  return storage.run(merged, fn);
}

module.exports = { current, community, requestId, traceId, bindRequest };

File: python/obs_sdk/_context.py
"""请求级通用字段的 context 读写（Python 版 sdkctx）。

用 contextvars 实现（线程 / async 各自隔离，与 spec/common-fields.md 一致）：
  - community 覆盖值 / request_id / trace_id 存于当前 Context；
  - 框架适配器在入口把可信解析出的字段 bind 进 Context，退出时 reset；
  - log / metrics 读取当前 Context，未绑定则回退部署级默认。
"""

from __future__ import annotations

import contextvars
import dataclasses
from contextlib import contextmanager
from typing import Iterator


@dataclasses.dataclass
class Request:
    """请求级通用字段。空串 / None 表示未设置。"""

    community: str | None = None
    request_id: str | None = None
    trace_id: str | None = None


_current: contextvars.ContextVar[Request] = contextvars.ContextVar(
    "obs_request", default=Request()
)


def get() -> Request:
    """读取当前请求级字段；未设置返回空 Request。"""
    return _current.get()


def community() -> str | None:
    """当前 community 覆盖值；未设置返回 None。"""
    return _current.get().community


def request_id() -> str | None:
    return _current.get().request_id


def trace_id() -> str | None:
    return _current.get().trace_id


@contextmanager
def bind(*, community: str | None = None, request_id: str | None = None,
         trace_id: str | None = None) -> Iterator[None]:
    """把请求级字段 bind 进当前 Context；退出自动 reset。

    用于框架适配器入口 / 业务可信判定点：

        with bind(community="openeuler", request_id=req_id):
            logger.info("...")   # 自动带 community/request_id
    """
    prev = _current.get()
    merged = Request(
        community=community if community is not None else prev.community,
        request_id=request_id if request_id is not None else prev.request_id,
        trace_id=trace_id if trace_id is not None else prev.trace_id,
    )
    token = _current.set(merged)
    try:
        yield
    finally:
        _current.reset(token)
